core/dbt/task/run.py

```diff
@@ -486,3 +489,52 @@ def get_runner_type(self, _):
     def task_end_messages(self, results):
         if results:
             print_run_end_messages(results)
+
+    def manage_schema(self, adapter, results: List[RunResult]):
+        # Read config
+        manage_schemas_config = self.config.manage_schemas  # type: bool
+        managed_schemas_actions_config: Dict[Tuple[str, str], str] = {
+            (ms.database or "", ms.schema or ""): ms.action or "warn"
+            for ms in self.config.managed_schemas
+        }
+
+        if not manage_schemas_config:
+            # TODO debug not doing anything
+            warn("Schema's configured to be managed, but manage_schemas is false in the profile")
+            return
+
+        if len(managed_schemas_actions_config) == 0:
+            warn_or_error("Schema management enabled for connection but no schema's configured to manage")
+            return
+
+        # Never manage schema if we have a failed node
+        was_successfull_complete_run = not any(
+            r.status in (NodeStatus.Error, NodeStatus.Fail, NodeStatus.Skipped) for r in results
+        )
+        if not was_successfull_complete_run and manage_schemas_config:
+            warn("One or more models failed, skipping schema management")
+            return
```

Comment:
Okay, this part is really interesting to me!

What's the rationale for skipping schema management if a model fails to build, a test finds failing records, etc?

I ask because, in seeing this code here, I have two thoughts:
- This feels more appropriate as a separate standalone task rather than a "hook" at the end of `run`, and its child tasks `seed`/`snapshot`/`build`. (What would we call that task? `dbt manage`? `dbt cleanup`... except for the similarity to `dbt clean`)
- It might be more appropriate to write this logic in a dbt-Jinja macro! That would then gives the user greater flexibility about how/when they run it. It could be a standalone `run-operation`, or an `on-run-end` hook

The main appeal of writing this logic in Python, and having it run here:
- Python gives us access to parallelism, if we wanted to run a bunch of `drop` statements concurrently
- Python gives us more direct access to the cache, in ways that we don't directly expose to the Jinja context — though Jinja can do relation-by-relation lookups, using `get_relation`
- We'd need to integrate it into the `run` task if we want to follow the logic here, skipping schema management if anything fails — though it would be possible to do the same in an `on-run-end` hook by [examining the `Results` object](https://docs.getdbt.com/reference/dbt-jinja-functions/on-run-end-context#results)

Reply:
@jtcohen6
- Moving to a macro is not a good option, see the original issue for that: a does not enjoy the dbt database driver abstraction that makes the logic portable across different databases. So it makes sense to have this as a core feature where it can be properly tested and maintained.
- Moving it into a separate command is definitely an option, and we could also drop the global config, making it easier to use. We could re-write the logic into a separate command called `dbt manage`.

However, would that make the change something the dbt-core team would likely accept? I would love to hear from you if the rest of your team agrees and would be willing to merge this if it were a separate command?

Reply:
> Moving to a macro is not a good option, see the original issue for that: a does not enjoy the dbt database driver abstraction that makes the logic portable across different databases.

Right, I couldn't remember if we had already discussed a hook/operation macro in the original issue.

I don't think I agree with your premise: a macro could enjoy the dbt database adapter abstraction, by tapping into methods that are (re)implemented across adapters as needed:
- listing all relations in a schema (i.e. `adapter.list_relations`) — that specific method isn't currently available to the Jinja context, but we could expose it, and/or use [other macros](https://github.com/dbt-labs/dbt-utils#get_relations_by_pattern-source) in the meantime
- dropping a relation / schema (i.e. `adapter.drop_relation`, `adapter.drop_schema`)

Some totally untested pseudo-code:
```
{% macro manage_schemas() %}

  {#-- pseudo code: let's say we have a reasonable way to access the config in the context --#}
  {% set manage_schemas_config = ... %}

  {% for database, schema, action in manage_schemas_config %}
  
    {% set actual_database = generate_database_name(database) %}
    {% set actual_schema = generate_schema_name(schema) %}
    
    {#-- pseudo code: find the database objects currently in this 'database.schema' --#}
    {#-- using a dbt_utils macro for this, but it could be a cache lookup via adapter.list_relations --#}
    {% set db_relations = dbt_utils.get_relations_by_pattern(
      schema_pattern = actual_schema,
      table_pattern = '%',
      exclude = '',
      database = actual_database
    ) %}
    
    {#-- pseudo code: use the 'graph' object to find currently enabled resources in this 'database.schema' --#}
    {% set enabled_resource_identifiers = graph.nodes.values()
                                          | selectattr("database", "equalto", actual_database)
                                          | selectattr("schema", "equalto", actual_schema)
                                          | map(attribute="identifier")
                                          | list %}

    {% for relation in db_relations %}
      {#-- pseudo code: for each real database object in this 'database.schema', see if it matches an enabled resource --#}
      {#-- if there's no match, warn, ignore, or drop the object --#}
      {% if relation.identifier not in enabled_resource_identifiers %}
        {% if action == 'drop' %}
          {{ log("Dropping " ~ relation) }}
          {{ adapter.drop_relation(relation) }}
        {% elif action == 'warn' %}
          {{ exceptions.warn("Found unknown object " ~ relation) }}
        {% elif action == 'ignore' %}
          {#-- noop --#}
        {% else %}
          {{ exceptions.warn("Unknown schema management action " ~ action) }}
        {% endif %}
    {% endfor %}

  {% endfor %}

{% endmacro %}
```

I agree that Python methods are much more easily unit tested than Jinja macros. Still, we'd have access to the same `pytest`-based framework which we currently use to test all runtime dbt functionality, including materializations, which necessarily combines logic written in both Python and Jinja.

So I really see the trade-off here as between:
- If written as a dbt-Jinja macro: Users can control when they run this (hook/operation), and can modify the logic if needed.
- If written as a Python task: Queries can be parallelized. The logic is simpler to write and test, but cannot be modified or customized by end users.

Reply:
Hi @jtcohen6 . This PR has been stale for a while due to the summer period. We now have time again to finish this up.

We think that a separate command `dbt manage` is the most maintainable solution to this problem. Is your team willing to accept the PR if we change the current logic to be a separate command? If so, we can adapt the PR and finalize this feature.

Reply:
@agoblet I'm aligned with a separate `dbt manage` command. If you want to keep working on this, I do see a path toward including this in `dbt-core`, given the interest demonstrated in Slack.

We're working now on refactoring our CLI. That's not a blocker, it just means that any net-new work in your PR around CLI subcommands + flags would need to be tracked for parity in that initiative.

I think there were some unresolved questions in the thread above about the best way to specify `SchemaManagementConfiguration`.

While it's out of scope for this PR, I think there's an opportunity here to turn "schemas" (and perhaps even logical "databases") into objects more fully managed by dbt. Think: applying grants and adding descriptions (= persisted as comments), in addition to creating if they don't yet exist (current behavior) and dropping orphaned objects (this proposal / PR). Again, out of scope for now — but I'd be interested in a configuration syntax that could leave room for that extension in the future.

Reply:
Started a discussion about the latter point: https://github.com/dbt-labs/dbt-core/discussions/5781

Again, I do _not_ view the full answer to that discussion as in scope for this PR, but I would be supportive of a syntax that's more naturally extensible to eventually meet those other needs

Reply:
@jtcohen6 @bneijt 

- moved the schema management logic to a separate `dbt manage` command
- created an enum for the action (`skip`, `warn` or `drop`)
- made the `SchemaManagementConfiguration` extensible by renaming the `action` field to `prune_models`, with an additional `skip` action for schemas that should not be pruned, but should be subject to other future management activities. It differs slightly from your proposition in #5781, the database field is flattened. 

What do you think?

Reply:
Also integrated the `generate_schema_name` macro